examples/erc20/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+
+#[derive(Debug, Eq, PartialEq, failure::Fail)]
+pub enum Error {
+    #[fail(display = "Unknown error occured.")]
+    Unknown,
+
+    #[fail(display = "Only existing admins can perform this operation.")]
+    AdminPrivilegesRequired,
+
+    #[fail(display = "Insuffient funds for transfer from {:?}.", address)]
+    InsufficientFunds { address: Address },
+
+    #[fail(display = "Address {:?} has no allowance from address {:?}.", from, to)]
+    NoAllowanceGiven { from: Address, to: Address },
+
+    #[fail(
+        display = "Transfer request {} exceeds allowance {}.",
+        amount, allowance
+    )]
+    RequestExceedsAllowance { amount: u64, allowance: u64 },
+}
+
+#[derive(Service, Default)]
+pub struct ERC20Token {
+    /// Total supply
+    total_supply: u64,
+
+    /// Owner
+    owner: Address,
+
+    /// Admins
+    admins: HashSet<Address>,
+
+    /// Account holders
+    accounts: HashMap<Address, u64>,
+
+    /// Allowances
+    allowed: HashMap<Address, HashMap<Address, u64>>,
+}
+
+// A Transfer event struct
+#[derive(Serialize, Deserialize, Clone, Debug, Default, Event)]
+pub struct Transfer {
+    #[indexed]
+    pub from: Address,
+    #[indexed]
+    pub to: Address,
+    #[indexed]
+    pub amount: u64,
+}
+
+// An Approval event struct
+#[derive(Serialize, Deserialize, Clone, Debug, Default, Event)]
+pub struct Approval {
+    #[indexed]
+    pub sender: Address,
+    #[indexed]
+    pub spender: Address,
+    #[indexed]
+    pub amount: u64,
+}
+
+impl ERC20Token {
+    /// Constructs a new `ERC20Token`
+    pub fn new(ctx: &Context, total_supply: u64) -> Result<Self> {
+        let owner = ctx.sender();
+        let mut admins = HashSet::new();
+        admins.insert(owner);
+        let mut accounts = HashMap::new();
+        accounts.insert(owner, total_supply);
+
+        Ok(Self {
+            total_supply,
+            owner,
+            admins,
+            accounts,
+            ..Default::default()
+        })
+    }
+
+    /// Get balance
+    pub fn balance_of(&mut self, ctx: &Context) -> Result<u64> {
+        let balance = match self.accounts.entry(ctx.sender()) {
+            Entry::Vacant(_ve) => 0u64,
+            Entry::Occupied(oe) => *oe.get(),
+        };
+        Ok(balance)
+    }
+
+    /// Get total supply
+    pub fn total_supply(&mut self, _ctx: &Context) -> Result<u64> {
+        Ok(self.total_supply)
+    }
+
+    /// Add admin
+    pub fn add_admin(&mut self, ctx: &Context, admin: Address) -> Result<()> {
+        if !self.admins.contains(&ctx.sender()) {
+            return Err(Error::AdminPrivilegesRequired);
+        }
+        self.admins.insert(admin);
+        Ok(())
+    }
+}
+
+// Helper methods
+
+/// Balance of
+fn _balance(accounts: &mut HashMap<Address, u64>, addr: Address) -> u64 {
+    match accounts.entry(addr) {
+        Entry::Vacant(_ve) => return 0u64,
+        Entry::Occupied(oe) => return *oe.get(),
+    }
+}
+
+/// transfer method
+fn _transfer(
```

Comment:
In Rust, `_*` usually signals that the variable is unused. I'd suggest moving the helper methods into an `impl ERC20` block and making them non-pub functions.

Reply:
Replaced with `do_transfer`.